phep-proposal.md

```diff
@@ -0,0 +1,119 @@
+```
+
+In these languages, such errors are HARD errors and correspond to a invalid numeric literal (not just a random illegal character).
+
+### Proposal for Pharo
+
+Instead of a formal language definition or a list on complex rules, the following single rule is considered: LEGAL=*`_` is legal only between 2 digits*.
+
+This means that leading, trailing and duplicated `_` are forbidden including around special characters `r`, `.` and `e`.
+
+* Examples of valid literals: `1`. `1_1`. `1_1.1_1`. `1_1r1_1.1_1e1_1`.
+* Examples of invalid literals: `_1`. `1_`. `1__1`. `1_.1`. `1_e1`.
+
+Here a discussion on  behaviors for Pharo are:
+
+1. Implements LEGAL in the NumberParser class but still follows the principle of the longest valid numeric token.
+
+   `1_1_a` will be parsed as `1_1` (longest valid literal according to LEGAL) followed by `_a` ie `11` (integer) and `_a` (unary message). No error or warning message.
+
+   This behavior is error prone since weird syntax will be unexpectedly and silently parsed as "something".
+   However, this is the closest thing to the current behavior according to the current implementation. (e.g. can you guess how `16rDeADMAN1` is currently parsed?)
+   Syntax highlighting (both number literal and unknown message) and fast runtime error is expected to help the programmer to deal with these cases.
+
+2. Implements LEGAL in the NumberParser class but introduce an alternative illegal numeric literal as possible token.
+
+  `1_1_a` will be parser as an error token `1_1_` ("numeric with trailing `_`") followed by the message send `a`.
+
+   This behavior is the closest to the other considered languages (Java, Python, Ruby).
+
+   Currently, `NumberParser`usually try to parses until it can't. In most case, the caller either accept the longest valid numeric literal, or complains if the whole input is not used.
+   The equivalent of "a bad formatted numeric token" is handled trough the method `NumberParser>>#expected:` that signals a runtime error or callback the client or UI (the ugliness of this design is out of the scope of the discussion). E.g. `NumberParser parse: '0r'`.
+
+3. Accept unrestricted use of `_` at the parsing level (no constraints at all) following the existing principle of the longest valid numeric token. LEGAL is applied at a second step at the Rule level (ie as a critique/warning of bad style).
+
+   For instance, it means that `1_1_a` will be parsed as `1_1_` (longest unrestricted numeric literal) followed by `a`.
+   ie. `11` (integer) and `a` (unary message send).
+   And that the editor/commit/etc. will complain about how ugly `1_1_` is.
+
+   The current PR advocates for this behavior since it (i) is the simplest thing to implement in `NumberParser`; (ii) move the responsibility of asserting ugliness and formatting to something more concerned about that (Rules, auto-formater, etc.); and (iii) this reduces the nitpicking and subtle differences between the specification of other languages. 
```

Comment:
```suggestion
   The current PR advocates for this behavior since it (i) is the simplest thing to implement in `NumberParser`; (ii) moves the responsibility of asserting ugliness and formatting to something more concerned about that (Rules, auto-formater, etc.); and (iii) this reduces the nitpicking and subtle differences between the specification of other languages. 
```